pkg/pool-stable/contracts/meta/MetaStablePool.sol

```diff
@@ -0,0 +1,372 @@
+        uint256[] memory balances,
+        uint256 lastChangeBlock,
+        uint256 protocolSwapFeePercentage,
+        uint256[] memory scalingFactors,
+        bytes memory userData
+    )
+        internal
+        virtual
+        override
+        returns (
+            uint256 bptAmountIn,
+            uint256[] memory amountsOut,
+            uint256[] memory dueProtocolFeeAmounts
+        )
+    {
+        // If the contract is paused, swap protocol fee amounts are not charged and the oracle is not updated
+        // to avoid extra calculations and reduce the potential for errors.
+        if (_isNotPaused()) {
+            _updateOracle(lastChangeBlock, balances[0], balances[1]);
+        }
+
+        return
+            super._onExitPool(
+                poolId,
+                sender,
+                recipient,
+                balances,
+                lastChangeBlock,
+                protocolSwapFeePercentage,
+                scalingFactors,
+                userData
+            );
+    }
+
+    // Oracle
+
+    function getOracleMiscData()
+        external
+        view
+        returns (
+            int256 logInvariant,
+            int256 logTotalSupply,
+            uint256 oracleSampleCreationTimestamp,
+            uint256 oracleIndex,
+            bool oracleEnabled
+        )
+    {
+        bytes32 miscData = _getMiscData();
+        logInvariant = miscData.logInvariant();
+        logTotalSupply = miscData.logTotalSupply();
+        oracleSampleCreationTimestamp = miscData.oracleSampleCreationTimestamp();
+        oracleIndex = miscData.oracleIndex();
+        oracleEnabled = miscData.oracleEnabled();
+    }
+
+    /**
+     * @dev Balancer Governance can always enable the Oracle, even if it was originally not enabled. This allows for
+     * Pools that unexpectedly drive much more volume and liquidity than expected to serve as Price Oracles.
+     *
+     * Note that the Oracle can only be enabled - it can never be disabled.
+     */
+    function enableOracle() external whenNotPaused authenticate {
+        _setOracleEnabled(true);
+
+        // Cache log invariant and supply only if the pool was initialized
+        if (totalSupply() > 0) {
+            _cacheInvariantAndSupply();
+        }
+    }
+
+    function _setOracleEnabled(bool enabled) internal {
+        _setMiscData(_getMiscData().setOracleEnabled(enabled));
+        emit OracleEnabledChanged(enabled);
+    }
+
+    function getLargestSafeQueryWindow() external pure override returns (uint256) {
+        return 34 hours;
+    }
+
+    function getLatest(Variable variable) external view override returns (uint256) {
+        int256 instantValue = _getInstantValue(variable, _getMiscData().oracleIndex());
+        return LogCompression.fromLowResLog(instantValue);
+    }
+
+    function getTimeWeightedAverage(OracleAverageQuery[] memory queries)
+        external
+        view
+        override
+        returns (uint256[] memory results)
+    {
```

Comment:
Does it make sense to try to extract these functions into a common module?

Reply:
Let's tackle it in a separate PR